src/data-model-providers/codegen/CodegenDataModelProvider.h

```diff
@@ -44,6 +44,9 @@ namespace app {
 class CodegenDataModelProvider : public DataModel::Provider
 {
 public:
+    // access to the typed global singleton of this class.
+    static CodegenDataModelProvider & Instance();
```

Comment:
So this is a little odd: the idea of having injectable providers is that the application knows about them and can register things with the provider.  But we're inverting that and having the cluster impl automagically register itself with the provider... except it's doing it from callbacks that are being called based on what, exactly?  I guess based on the endpoint config registered via the ember stuff?

I'm not really happy with making these singleton assumptions, but I guess they're true in practice and I don't have any better ideas....

Reply:
The reason for this is that for codegen integration we call `Register` on the codegen registry. That is specific to the `codegen provider` so the integration call needs to get a `codegen instange` rather than a generic provider instance.

The way I see the path forward is that we are looking to have our application `main` register things when we are fully code driven. However for codegen instances we have to do things through code (unless/until we update things to codegen different things .. and even then we need dynamic cluster support).